### DISCUSSION TITLE
getenv vs $_ENV mismatch

### DISCUSSION THREAD

**Discussion by ahundiak** [symfony/symfony#49928] [Q&A] [answered]
So with a fresh 6.2 install, a default make:controller and using symfony server:start I noticed that `dump(getenv())` (in the controller action) shows a bunch of system variables but nothing from .env.  Not even APP_ENV.  On the other hand dumping $_ENV shows them all.

I**s this mismatch intentional?**  I know you should not be using getenv (or $_ENV) but I would still expect them to be the same.  I also know there is processing going on so only variables actually being used end up in in the container and what not but I'm pretty sure the default variables in the delivered .env file are being used.

I did a bit of searching but as you might imagine getenv generates quite a few hits.  Nothing really jumped out.

**Comment by ghost** [comment]
By default Symfony doesn't use putenv (see stof comment to know why). See https://github.com/symfony/symfony/blob/v6.2.6/src/Symfony/Component/Runtime/SymfonyRuntime.php#L106

**Reply by stof** [reply, marked as answer]
The reason to avoid using `putenv` is that this method is not thread-safe (and if you use Apache's mod-php to run PHP, it relies on threads)

**Reply by ghost** [reply]
> I also know there is processing going on so only variables actually being used end up in in the container and what not but I'm pretty sure the default variables in the delivered .env file are being used.

Actually getenv/$_ENV has nothing to do with Symfony container. Symfony container only uses a `%env()%` pseudoparameter syntax. It also doesn't check during compilation whether env var exists (it will throw on runtime if not). The whole purpose of env vars is the ability to modify application behaviour on runtime. I really recommend Nicolas Grekas talk on configuring Symfony where he explains everything in details https://symfonycasts.com/screencast/symfonycon2019/configuring-symfony-from-localhost-to-high-availability

**Reply by stof** [reply]
@javaDeveloperKid you are confusing the `%env()%` feature of `symfony/dependency-injection` and the `symfony/dotenv` component.

**Reply by ghost** [reply]
@stof as I understand the quoted sentence, the OP thinks that the fact the container is using env vars (through `%env()%` pseudo parameter) means they exist in $_ENV and getenv. It's not true. It's the symfony/dotenv that populates $_ENV and getenv, not the container (symfony/dependency-injection) that only reads them.

**Reply by ahundiak** [reply]
Just to clarify, what originally kicked my research off was a question which wondered why `bin/console dump:container --env-vars` seemed to show some variables in .env but not the ones the original poster had added.  We explained that `--env-vars` only showed variables being used.  Eventually we coaxed the OP into showing some code and it turns out there were using static functions and getenv. Legacy stuff.

Hence my original remark about the container.  Which in retrospect probably just confused the issue.  But at the time I was thinking that maybe there was a link.  The runtime component had never even crossed my mind at that point.

**Comment by ahundiak** [comment]
Just wanted to say thanks for the comments.  Now I got something I can link to when the question comes up again.  Still think it is strange that there is a mismatch between $_ENV and getenv but that is okay.  Lots of things I don't understand and it's clearly not been a major issue

**Reply by stof** [reply]
The Dotenv component _can_ be configured to use `putenv` as well to update the value returned by `getenv`. But it is disabled by default due to the thread safety issue.

**Reply by ahundiak** [reply]
Can I get a hint (or perhaps a link) as to how to do this configuration within the framework?  Not seeing anything under the framework configuration.  Maybe in the runtime component?

**Reply by stof** [reply]
in the fullstack framework, the DotEnv component is generally configured through the runtime component.

This is not in the FrameworkBundle configuration because DotEnv must be used _before_ starting to boot the kernel.

**Reply by ahundiak** [reply]
Thanks. This did the trick:
```
#composer.json
    "extra": {
        "runtime": {
            "use_putenv": true
        },
```

**Reply by ghost** [reply]
> Thanks. This did the trick:
> 
> ```
> #composer.json
>     "extra": {
>         "runtime": {
>             "use_putenv": true
>         },
> ```

I didn't know about this way. Another way is to use APP_RUNTIME_OPTIONS env var
